File: Python/Nonlineq/SecantBisNewton_convergence.py

import numpy as np #Basic linear algebra
import matplotlib.pyplot as plt #Plotting


tol = 10*10**(-15)
def f(x):
    return x**10 - 10**10 
def f_prime(x):
    return 10*x**9
#%%
#Halveringsmetoden

n=0 #intervall
p=10**10
halvering_x=[]
halvering=[]
k=0
while abs(n-p)>tol:
    m = (n+p)/2
    if f(m)<0:
        n=m
    else:
        p=m
    halvering_x.append(p)
    halvering.append(f(p))
    k += 1

i=np.linspace(1,k,k) #stara från 1 iteration
plt.plot(i,halvering,"r.-",label="Halveringsmetoden")
plt.legend()
#%%
x0 = 5
x1 = 7
sekant_x=[]
sekant=[]
sekant_x.append(x0);sekant_x.append(x1)
sekant.append(f(x0));sekant.append(f(x1))
k2 = 2 #gissningar räknas som iteration
while abs(x1-x0)>tol:
    x = x1-f(x1)*((x0-x1)/(f(x0)-f(x1)))
    x0 = x1
    x1 = x
    sekant_x.append(x)
    sekant.append(f(x)) #spara f(x) i array för senare plot
    k2 += 1
i2=np.linspace(1,k2,k2) #stara från 1 iteration
plt.plot(i2,sekant,"bo-",label="Sekantmetoden") #plotta antal iteration och f(x)
plt.legend()
#%%

y0=10**10 #första gissning 
y=y0-f(y0)/f_prime(y0)
newton_x=[]
newton=[]
newton_x.append(y0); newton_x.append(y)
newton.append(f(y0)); newton.append(f(y))
k3=2

while abs(y-y0)>tol:
    y0=y
    y=y0-f(y0)/f_prime(y0)
    newton_x.append(y)
    newton.append(f(y))
    k3 += 1

i3=np.linspace(1,k3,k3) #stara från 1 iteration
plt.plot(i3,newton,"c*-",label="Newtons metod")
plt.legend()

plt.xlabel('iteration')
plt.ylabel('funktion x**10 - 10**10')

plt.title('Sekant, halvering  och Newton  för x(1)=' f'{sekant_x[0]}' ',x(2)=' f'{sekant_x[1]}')

File: MATLAB/Nonlineq/SecantBisNewton_convergence.m
close all;

tolerance = 10e-10;

fun = @(x)x^10 - 10^10;

% definition av intervalet [n,p]

n= 0;
p=10^10;
it=0;

% Halveringsmetoden

	  while abs(n - p) > tolerance
m = (n + p)/2;
         if fun(m) < 0 
  n = m;
         else
	   p = m;
        end

	it = it+1;

func_val(it) =fun(p);
iteration(it) = it;
end

%Sekantmetoden

x=0;
iteration_sekant = 0;
func_sekant = 0;
k=2;
%Initialiseringen
x(1) = 5;
x(2) = 7;


iteration_sekant(1) = 1;
iteration_sekant(2) = 2;
func_sekant(1) =fun(x(1));
func_sekant(2) =fun(x(2));


while  abs(x(k) - x(k-1)) > tolerance

numerator =  fun(x(k))*(x(k-1) - x(k));
denominator = fun(x(k-1)) - fun(x(k));

x(k+1) = x(k) - numerator/denominator;

k = k+1;
  iteration_sekant(k) = k;
func_sekant(k) =fun(x(k));


end

	
figure

plot(iteration,  func_val,'o r-',  'LineWidth',2)

  hold on
plot(iteration_sekant,func_sekant, 'o b-',  'LineWidth',2)


% Newtons metod

y=0;
iteration_newton = 0;
func_newton = 0;


%Initialiseringen
y(1) = 10^10;
k=2;

iteration_newton(1) = 1;
func_newton(1) =fun(y(1));
numerator =  fun(y(1));
denominator = 10*y(1)^9;
y(2) = y(1) - numerator/denominator;
  
% Main Newton's iterations  
while abs( y(k) - y(k-1) ) > tolerance

numerator =  fun(y(k));
denominator = 10*y(k)^9;
y(k+1) = y(k) - numerator/denominator;
iteration_newton(k) = k;
func_newton(k) =fun(y(k));
k = k+1;
  end

  plot(iteration_newton,func_newton, 'o c-','LineWidth',2)

  
  xlabel('iteration')
  ylabel('funktion x^{10} - 10^{10}')
  legend('Halveringsmetoden','Sekantmetoden','Newtons metod')
  title(['Sekant, halvering  och Newton  för x(1)=',num2str(x(1)),', x(2)= ',num2str(x(2))])

iteration

 iteration_sekant

 iteration_newton
